src/msgtree.rs

```diff
@@ -0,0 +1,178 @@
+                (false, &TreeGraphElem::Empty    ) => ' ',
+                (false, &TreeGraphElem::Following) => '|',
+            })?;
+        }
+        Ok(())
+    }
+}
+
+
+/// Iterator returning the graph elements for a single commit
+///
+/// Only the first line returned by the iterator will contain marks.
+///
+pub struct CommitTreeGraphLines(Vec<TreeGraphElem>);
+
+impl Iterator for CommitTreeGraphLines {
+    type Item = TreeGraphElemLine;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let retval = TreeGraphElemLine(self.0.clone());
+
+        // marks shall only be part in the first line for each commit
+        for elem in self.0.iter_mut() {
+            match elem {
+                &mut TreeGraphElem::Mark(MarkType::End) => *elem = TreeGraphElem::Empty,
+                &mut TreeGraphElem::Mark(_) => *elem = TreeGraphElem::Following,
+                _ => {},
+            }
+        }
+
+        Some(retval)
+    }
+}
+
+
+/// Iterator generating graph elements for a series of commits
+///
+pub struct TreeGraphElemLineIterator<'r, I>
+    where I: Iterator<Item = Commit<'r>>
+{
+    inner: I, // inner iterator over commits for which to display the graph
+    parents: Vec<Option<Oid>>, // currently tracked parents
+}
+
+impl<'r, I> Iterator for TreeGraphElemLineIterator<'r, I>
+    where I: Iterator<Item = Commit<'r>>
+{
+    type Item = (TreeGraphElemLine, Commit<'r>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().map(|commit| {
+            // We will definitely require the parent id in order to draw a graph.
+            // However, we only want to track the parent once, so we end up with
+            // nice horizontal merges in our graph.
+            let mut parent_update = commit.parent(0).as_ref().map(Commit::id).ok();
+
+            // generate graph elements for the parents currently tracked
+            let mut elems : TreeGraphElemLine = self.parents.iter_mut().map(
+                |mut parent| match *parent {
+                    Some(id) => if commit.id() == id {
+                            // the current commit is a parent we were awaiting
+                            *parent = parent_update.take();
+                            TreeGraphElem::Mark(if parent.is_some() {MarkType::Mid} else {MarkType::End})
+                        } else { TreeGraphElem::Following },
+                    None => TreeGraphElem::Empty,
+                }
+            ).collect();
+
+            if parent_update.is_some() {
+                // a history for the current commit doesn't yet exist, so we
+                // should create one
+                elems.append(TreeGraphElem::Mark(MarkType::Start));
+                self.parents.push(parent_update);
+            }
+
+            // keep the graph slim by removing parent trackig information that
+            // is no longer required from the back of the list
+            while self.parents.last().map(|p| p.is_none()).unwrap_or(false) {
```

Comment:
Simple way to improve readability here: `.map(Option::is_none).unwrap_or(false)`!